src/bin/pg_autoctl/pgsql.c

```diff
@@ -165,6 +205,116 @@ pgsql_open_connection(PGSQL *pgsql)
 }
 
 
+/*
+ * pgsql_retry_open_connection loops over a PQping call until the remote server
+ * is ready to accept connections, and then connects to it and returns true
+ * when it could connect, false otherwise.
+ */
+static PGconn *
+pgsql_retry_open_connection(PGSQL *pgsql)
+{
+	PGconn *connection = NULL;
+	int attempts = 0;
+	bool retry = true;
+	bool connectionOk = false;
+	PGPing rv;
+	uint64_t startTime = time(NULL);
+
+	log_warn("Failed to connect to \"%s\", retrying until "
+			 "the server is ready", pgsql->connectionString);
+
+	while (retry)
+	{
+		uint64_t now = time(NULL);
+
+		if ((now - startTime) >= POSTGRES_PING_RETRY_TIMEOUT)
```

Comment:
does this needs to be configurable. 15 min looks little long to me

Reply:
The expectation is that we face a transient error. We don't want to fail too soon, and I agree that 15 mins is a long time. For interactive use-cases, it's still possible to interrupt (^C) the loop easily, though. Also, implementing configuration parameters down in `pgsql.c` would require a lot of changes, and I don't think it's important enough to be able to change the default here.

Another (cheap) way to do it would be via an environment variable, maybe. I'm just not sure it's worth it.

Reply:
I would rather have this way than having a magic env variable.